src/interfaces/messages/types.ts

```diff
@@ -0,0 +1,24 @@
+import type { BaseMessageReply } from '../../core/message-reply.js';
+import type { BaseMessage, Descriptor } from '../../core/types.js';
+import type { DwnInterfaceName, DwnMethodName } from '../../core/message.js';
+
+export type MessagesGetDescriptor = Pick<Descriptor, 'interface' | 'method' > & {
+  interface : DwnInterfaceName.Messages;
+  method: DwnMethodName.Get;
+  messageCids: string[];
+};
+
+export type MessagesGetMessage = BaseMessage & {
+  descriptor: MessagesGetDescriptor;
+};
+
+// oof horrible name
+export type MessagesGetReplyEntry = {
+  messageCid: string;
```

Comment:
i wonder if we should return a `Map`/`Object` where the keys are the `messageCid` and the values are the `{ message, encodedData }` so that the caller can:

- ensure they will never encounter the same `messageCid` more than once (since keys are unique)
- quickly grab the `message`/`encodedData` for a given `messageCid` instead of having to iterate the entire array to find the entry matching the `messageCid`
- possibly even avoid storing the `messageCid` (at least without having to manipulate the data) if they wanted to do something like `savedMessages.push(Object.values(result))`

Reply:
> possibly even avoid storing the `messageCid` (at least without having to manipulate the data) if they wanted to do something like `savedMessages.push(Object.values(result))`

Not sure what you mean by this

> quickly grab the message/encodedData for a given messageCid instead of having to iterate the entire array to find the entry matching the messageCid

In practice, I think clients will generally be iterating over the array so they can process messages in order, but you're right that it may benefit the odd client. If we use an object with `messageCid`s as keys, we don't return an ordered list. [Looks like some JSON serializers don't preserve key order](https://filemakerhacks.com/2020/04/29/ordering-elements-within-json-objects/#:~:text=The%20JSON%20Data%20Interchange%20Standard,%2C%20and%20should%20not%2C%20matter.) That might not be a big deal; the client could remember the list of `messageCids` they sent in the request and use that to determine the order of messages to process. Not arguing against your approach, just something to think about as we put sync in production.